### ISSUE TITLE
Fixing RPC method resolution for python 3

### ISSUE THREAD

**Issue by leifurhauks** [grpc/grpc#5917]
An obstacle still in the way of python 3 compatibility is that rpc methods don't get resolved.

Working in [my own topic branch](https://github.com/leifurhauks/grpc/tree/topic/py3_compatibility) that includes #5660, #5657, and #5368 (still being reviewed), many of the test failures look like the following:

```
_intermediary_low_test.EchoTest.testManyOneByteEchoes
AssertionError: 'test method' != b'test method'
```
(I omitted the traceback for clarity. The full traceback is [here](https://gist.github.com/leifurhauks/63c2e19caee35f0b6179))

What happened here is that the server's rpc methods are indexed by unicode string keys, and byte strings are being used for lookup.

I believe this is why there are a lot of failures that look like this one:
```
_crust_over_core_over_links_face_interface_test.GenericInvokerBlockingInvocationInlineServiceTest.testSuccessfulUnaryRequestUnaryResponse

grpc.framework.interfaces.face.face.RemoteError: RemoteError(code=StatusCode.UNKNOWN, details="b''")
```
([full traceback](https://gist.github.com/leifurhauks/4bf01675ebab2f85c6a2))

I think that when requests are received from the wire, the method names are deserialized as byte strings.

I experimented by changing the python protoc plugin to build the method and serializer dictionaries with byte string keys, instead of string literals (which are unicode strings in python 3). After fixing a string issue in `grpc._links.service` (see #5916) I was actually able to run the helloworld and route_guide python examples without errors on python3.4 (though there were still a lot of test failures).

For reference, I posted the [route_guide_pb2.py](https://gist.github.com/leifurhauks/30924d532baa2d707bc5#file-route_guide_pb2-py-L318) generated after I modified the protoc plugin. 

Is this the right way to deal with rpc method resolution on python 3? Or should the method names extracted from each request be decoded into unicode strings for the lookup?

CC: @nathanielmanistaatgoogle @soltanmm

**Comment by leifurhauks** [comment]
re: #282

**Comment by soltanmm** [comment]
Anything that is a name needs to be ASCII or encoded as far as I'm aware, so that seems right.

**Comment by nathanielmanistaatgoogle** [comment]
You're right that method names need to be bytes when they are passed from Python to the C core and you're right that the C core passes method names as bytes to Python on the service side. However, the developer experience at the application level needs to be method names as "native strings" - so that means `bytes` in Python 2 and unicode (`str`) in Python 3. So at a point of our choosing between the two we need to be doing the detection and translation. I hope it's easy because I don't believe that we need to be accommodating - if an application passes us a `unicode` method name in Python 2 or a `bytes` method name in Python 3 then we can just raise a `TypeError`.

Hmm... _should_ we be at all accommodating? Would it be helpful of us if we were?

**Comment by leifurhauks** [comment]
I'm not sure I understand the question. I don't think the application developer would need to handle the method name strings directly, although they would be visible as the keys to `my_server_instance._implementations`, after getting a server instance from the generated server factory.

If I understand correctly, the only way an app developer would be able to pass the wrong type of string would be if they edited the generated pb2 file directly, changing the server factory.

Using the solution I experimented with, I was able to run the examples without modifying the concrete servicers (or any example code, beyond changing print statements to print functions).

However I've just realized a problem with the way I did it -- the `_b` function in the pb2 file encodes python 3 strings as `latin1`, this is because it's used to wrap a string literal containing the binary serialized `.proto` file. But I think method names should be encoded as `utf8`, because method names in the `.proto` file should be protobuf strings: https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto#L234

**Comment by nathanielmanistaatgoogle** [comment]
Everything you point out is correct. I think I got ahead of myself looking forward to a day in which stub objects support dynamic discovery of methods.

**Comment by nathanielmanistaatgoogle** [comment]
Having thought about this some more: we have an explicit goal to keep the generated code as simple as possible, and doing string type conversion in generated code is something we explicitly want to avoid. So given that the surface of the generated code needs to work in terms of native strings, that means the surface of the non-generated code (`grpc.beta`, `grpc.framework`, and friends) needs to work in terms of native strings as well. Since the layers under the non-generated (static) runtime code need to work in terms of bytes, that means that the conversion should be done somewhere in the static runtime code.

Does that make sense?

**Comment by leifurhauks** [comment]
Yeah, I think that makes sense. Are you suggesting that the conversion should occur at the interface with the compiled extension, or somewhere in the layers above that?

**Comment by nathanielmanistaatgoogle** [comment]
The conversion should occur probably just above the Cython layer.

The many, many layers above that were built with goals in mind that we've mostly abandoned and now serve more than anything else to slow the system down. They're going to be mostly excised as soon as can be managed.

**Comment by nathanielmanistaatgoogle** [comment]
@kpayson64: please reassess in light of recent work?

**Comment by kpayson64** [comment]
This was resolved by #6503.
